### ISSUE TITLE
latest miniconda breaks gcc

### ISSUE THREAD

**Issue by dnadeau4** [ContinuumIO/anaconda-issues#6340]
New miniconda did not pass `-Wl,--enable-new-dtags,-R/home/travis/miniconda/lib` to the linker.  I reverted from latest to version `Miniconda3-4.3.21-Linux-x86_64.sh` and `python setup.py install` passed the parameter correctly to gcc

Here is the error `-Wl,--enable-new-dtags` is missing and there is a new `-B /home/travis/miniconda/compiler_compat` parameter.

 * gcc -pthread -shared `-B /home/travis/miniconda/compiler_compat` -L/home/travis/miniconda/lib -Wl,-rpath=/home/travis/miniconda/lib,--no-as-needed -w -g -O0 build/temp.linux-x86_64-3.6/Src/Cdunifmodule.o -L/home/travis/miniconda/lib -L/home/travis/miniconda/lib -L/home/travis/miniconda/lib -L/usr/X11/lib -L/usr/X11R6/lib -L/home/travis/miniconda/lib `-R/home/travis/miniconda/lib`-lcdms -lnetcdf -lnetcdf -ldrsfortran -lgfortran -lgrib2c -lpng -ljasper -lpython3.6m -o build/lib.linux-x86_64-3.6/cdms2/Cdunif.cpython-36m-x86_64-linux-gnu.so`

`gcc: error: unrecognized command line option -R`
`gcc: error: unrecognized command line option -R`

**Comment by dnadeau4** [comment]
`wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh `

**Comment by mingwandroid** [comment]
Thanks for the report, I'll get round to this ASAP.

**Comment by mingwandroid** [comment]
Can you share your build script please?

**Comment by msarahan** [comment]
We have no idea where your -R arguments are coming from.  That's a linker argument, and it should not be passed to gcc.

The -B path is something we needed to add to use our newer ld executable when linking code with our libraries.  We observed this need particularly with numpy's static library, when building statsmodels.  This is part of the _sysconfigdata.py file in Python, which we've copied from older python packages to retrieve their flags.  This is where we've added the -B flag.  There is a separate _sysconfigdata.py file that is specific to our new tools that can be enabled with the _PYTHON_SYSCONFIGDATA_NAME environment variable.

The real "right" thing here is to either use our new compilers and binutils, or stick to older (<5) anaconda or <4.3.27 miniconda.  We are trying hard to support people who don't want to do that, but we're honestly not sure how far we'll be able to get.  We don't have very good documentation on how to use the new tools yet, but we hope to have that up very soon.  In the meantime, if you have advice for us on how we can better support backward compatibility with other build tools, please let us know.  We believe that our backward compatibility at runtime is already excellent - back to CentOS 6 on Linux, and OSX 10.9 - so there's little reason to keep using old toolchains.  We do have to ship libstdc++, since we're now on the new ABI - that may be the only reason I know of to stay with the older toolchains.

**Comment by sdrobert** [comment]
Thanks for your consideration. I think that Conda has wonderful support on a variety of platforms. That's why I use it :)

Do you mean just compiling with `binutils_linux-64` and `gcc_linux-64`? Including only the former in the conda build results in the same `-R` problem, and including both leads to

```
/u/sdrobert/.local-spock1/miniconda3/envs/foo/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -pipe -D_FORTIFY_SOURCE=2 -O2 -fPIC -DKALDI_DOUBLEPRECISION=0 -DHAVE_EXECINFO_H=1 -DHAVE_CXXABI_H -DHAVE_OPENBLAS -I/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src -I/u/sdrobert/.local-spock1/miniconda3/envs/foo/include -I/u/sdrobert/.local-spock1/miniconda3/envs/foo/include/python2.7 -I/u/sdrobert/.local/lib/python2.7/site-packages/numpy/core/include -c /ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/swig/pydrobert/kaldi_wrap.cpp -o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/swig/pydrobert/kaldi_wrap.o -std=c++11 -m64 -msse -msse2 -pthread -fPIC
x86_64-conda_cos6-linux-gnu-cc: error trying to exec 'cc1plus': execvp: No such file or directory
error: command '/u/sdrobert/.local-spock1/miniconda3/envs/foo/bin/x86_64-conda_cos6-linux-gnu-cc' failed with exit status 1
```

It's a bit tricky migrating to the recommended system as I can't seem to find much in the way of a tutorial. There's a bit [here](https://conda.io/docs/user-guide/tutorials/build-postgis.html), but it doesn't use gcc AFAICT.

Best,
Sean

**Comment by msarahan** [comment]
In addition to that blog post, were working on more docs specific to dealing with compilers. Draft at https://docs.google.com/document/d/103QY204eD82fJzpT7eXy_rV2bKwe6Hx5uAnkMg_xwkI/edit?usp=sharing

**Comment by sdrobert** [comment]
Egh, seems to be happening using conda's build 3 templates:

Updated meta.yaml:

``` yaml

```
gives
```
/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/bin/x86_64-conda_cos6-linux-gnu-c++ -pthread -shared -B /u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/compiler_compat -L/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -Wl,-rpath=/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib,--no-as-needed -Wl,-O2,--sort-common,--as-needed,-z,relro,-z,now -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -pipe -D_FORTIFY_SOURCE=2 -O2 build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/swig/pydrobert/kaldi_wrap.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/base/io-funcs.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/base/kaldi-error.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/base/kaldi-math.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/base/kaldi-utils.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/feature-fbank.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/feature-functions.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/feature-mfcc.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/feature-plp.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/feature-spectrogram.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/feature-window.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/mel-computations.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/online-feature.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/pitch-functions.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/resample.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/signal.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/feat/wave-reader.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/compressed-matrix.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/kaldi-gpsr.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/kaldi-matrix.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/kaldi-vector.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/matrix-functions.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/optimization.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/packed-matrix.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/qr.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/sp-matrix.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/sparse-matrix.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/srfft.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/matrix/tp-matrix.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/thread/kaldi-barrier.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/thread/kaldi-mutex.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/thread/kaldi-semaphore.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/thread/kaldi-thread.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/transform/cmvn.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/kaldi-holder.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/kaldi-io.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/kaldi-table.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/parse-options.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/simple-io-funcs.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/simple-options.o build/temp.linux-x86_64-2.7/ais/clspace5/u/sdrobert/scratch/pydrobert-kaldi/src/util/text-utils.o -L/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib64 -L/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -L/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -R/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib64 -R/u/sdrobert/.local-spock1/miniconda3/conda-bld/pydrobert-kaldi-openblas_1507081013650/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -lpthread -lm -ldl -lgfortran -lopenblas -lpython2.7 -o build/lib.linux-x86_64-2.7/pydrobert/kaldi/_internal.so
x86_64-conda_cos6-linux-gnu-c++: error: unrecognized command line option '-R'
x86_64-conda_cos6-linux-gnu-c++: error: unrecognized command line option '-R'
```
Conda info:

```
conda info
Current conda install:

               platform : linux-64
          conda version : 4.3.27
       conda is private : False
      conda-env version : 4.3.27
    conda-build version : 3.0.23
         python version : 3.6.2.final.0
       requests version : 2.18.4
```

**Comment by msarahan** [comment]
Did you set _PYTHON_SYSCONFIGDATA_NAME, as detailed in https://github.com/conda/conda/issues/6030#issuecomment-333271364?

This is an annoying necessity right now.  The compiler activation scripts should do it for you, but we need to rebuild them to add that.  WIP.  For now, adding it in a conda_build_config.yaml file should work.  That file can be alongside meta.yaml, or it can be many other places.  The search order is described at https://conda.io/docs/user-guide/tasks/build-packages/variants.html#creating-conda-build-variant-config-files

The -R errors you're seeing generally indicate that Python can't figure out what kind of compiler it's using.  It points to some issue with _sysconfigdata - the bridge between the compiler configuration and python.  Unfortunately, python does not seem to just check for CC and friends - it requires this extra dance.  The correct _sysconfigdata is there, but we don't enable it by default, out of respect for people who don't want to opt into using our new compilers yet.

Basically, -R is a flag for AIX or solaris, and they're the last in a long chain of things that python tries.

**Comment by msarahan** [comment]
I think this other issue finally gets to the bottom of it: https://github.com/ContinuumIO/anaconda-issues/issues/6410#event-1279078358

**Comment by sdrobert** [comment]
Good to hear. I should say that, for the time being, setting _PYTHON_SYSCONFIGDATA_NAME worked for linux (thank you!).

There are a few problems with the `openblas` recipe, `libgfortran`, and the like, but I'll wait to open issues until after the compiler stuff has settled down :)

**Comment by cvanelteren** [comment]
I am on miniconda 4.10.3. After my system updated to gcc11 i cannot compile anymore with cython. Is there a way to use the system gcc in miniconda? More info  https://github.com/cython/cython/issues/4218

**Comment by joshua-cogliati-inl** [comment]
This is possibly a related bug (also caused by gcc 11 and conda): https://github.com/stan-dev/pystan/issues/294

That bug is because GLIBCXX_3.4.29 added _ZSt28__throw_bad_array_new_lengthv and that symbol does not appear in conda's symbol libstdc++.so
